designs/validations.md

```diff
@@ -0,0 +1,288 @@
+	* Most eks-a API objects implement the `SetDefaults` method.
+	* Most webhooks are also implemented in this package. They are composed of a combination of `object.Validate` calls and field immutability checks (for `ValidateUpgrade`). For defaults, they call `SetDefaults`.
+* `pkg/cluster`
+	* This package exports `cluster.Config` and allows to set defaults and validate it with `cluster.SetDefaults` and `cluster.Validate`. These internally call a `ConfigManager`, which allows to register logic dynamically to parse from yaml, set defaults and validate a `cluster.Config`.
+	* Validations in the package either call `.Validate` in the API objects or implement static validations but at the `Config` level (eg. validations that need multiple API objects).
+	* `ConfigManager`'s defaults also call the `SetDefaults` method in API objects.
+* `pkg/gitops/flux`
+	* Exports validations through `Validations` method, which returns a slice of `validations.Validation`, hence designed to be used in conjunction with `validations.Runner`, etc. These are run from the create cluster workflow.
+* `pkg/providers`
+	* All providers implement `SetupAndValidateCreateCluster` and `SetupAndValidateUpgradeCluster`. They are called from the *create* and *upgrade* workflows.
+	* These methods mix API defaults, validations and internal state changes (equivalent to an `Init` method).
+	* The internal implementation of these method variates from provider to provider. Some of them have a `Validator` and a `Defaulter` and Snow, for example, uses the `cluster.ConfigManager` to organize defaults and validations.
+* `pkg/workflows`
+	* Compose and run validations and defaults with `validations.Runner` from a workflow task.
+* `test/framework`
+	* `ClusterValidator` is yet another validator runner. This is one centered in validating the state of a cluster against a given cluster specification by using a k8s client to inspect the cluster API.
+* Special note is due for commands in `cmd`, since these are the entities with more knowledge and they are a good example of the fragmentation addressed in this doc. Currently, they handle different types of validations and at different stages:
+	* Independently run validations, like:
+		* `FileExists` for the cluster config yaml.
+		* Parsing and Cluster config API validation (`GetAndValidateClusterConfig`) 
+		* Tinkerbell CLI flags.
+		* Registry mirror auth.
+		* Validations to `cluster.Config` with `cluster.Validate`. Defaults are applied by just calling `cluster.NewSpecFromClusterConfig`.
+	* "Grouped" validations in `commonValidations`
+		* Docker daemon config
+		* File exists, parsing and Cluster config API validation (`GetAndValidateClusterConfig`). This is duplicated.
+		* Kubeconfig file and format validation.
+	* Delegated validations to the workflows through `interfaces.Validator`.
+
+## Goals
+* Define a standard for types of validations/defaults and where and **where not** to put them.
```

Comment:
We might want to replace the `/` in some of these goals to make clear we're addressing 2 independent concepts, not a pseudonym as `/` can represents.

E.g.

```
* Define a standard for types of validation and defaults.
* Define where validations and defaults should be defined within the code base.
```